File: frontend/src/services/adminService.js
import { API_ENDPOINTS } from '../config/api';
import { authService } from './authService';

/**
 * Service handling administrator API operations.
 */
export const adminService = {
  /**
   * Fetches aggregate system metrics.
   */
  async getStats() {
    const res = await fetch(API_ENDPOINTS.ADMIN_STATS, {
      headers: authService.getAuthHeaders(),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to fetch admin stats.');
    }
    return res.json();
  },

  /**
   * Fetches system users with optional search and role filter.
   */
  async getUsers(params = {}) {
    const url = new URL(API_ENDPOINTS.ADMIN_USERS);
    if (params.query) url.searchParams.set('query', params.query);
    if (params.role) url.searchParams.set('role', params.role);

    const res = await fetch(url.toString(), {
      headers: authService.getAuthHeaders(),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to fetch users.');
    }
    return res.json();
  },

  /**
   * Provisions a new user.
   */
  async createUser(userData) {
    const res = await fetch(API_ENDPOINTS.ADMIN_USERS, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...authService.getAuthHeaders(),
      },
      body: JSON.stringify(userData),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to create user.');
    }
    return res.json();
  },

  /**
   * Updates a user's system role.
   */
  async updateRole(userId, role) {
    const res = await fetch(API_ENDPOINTS.ADMIN_USER_ROLE(userId), {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        ...authService.getAuthHeaders(),
      },
      body: JSON.stringify({ role }),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to update user role.');
    }
    return res.json();
  },

  /**
   * Updates a user's enabled status.
   */
  async updateStatus(userId, enabled) {
    const res = await fetch(API_ENDPOINTS.ADMIN_USER_STATUS(userId), {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        ...authService.getAuthHeaders(),
      },
      body: JSON.stringify({ enabled }),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to update user status.');
    }
    return res.json();
  },

  /**
   * Fetches all submissions with optional filters.
   */
  async getSubmissions(params = {}) {
    const url = new URL(API_ENDPOINTS.ADMIN_SUBMISSIONS);
    if (params.week) url.searchParams.set('week', params.week);
    if (params.year) url.searchParams.set('year', params.year);
    if (params.section) url.searchParams.set('section', params.section);
    if (params.studentId) url.searchParams.set('studentId', params.studentId);

    const res = await fetch(url.toString(), {
      headers: authService.getAuthHeaders(),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to fetch submissions.');
    }
    return res.json();
  },

  /**
   * Deletes a submission and purges its stored files.
   */
  async deleteSubmission(id) {
    const res = await fetch(API_ENDPOINTS.ADMIN_SUBMISSION_DELETE(id), {
      method: 'DELETE',
      headers: authService.getAuthHeaders(),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || 'Failed to delete submission.');
    }
    return res.json();
  },
};

File: authportal/src/main/java/com/selva/authportal/service/AdminService.java
package com.selva.authportal.service;

import com.selva.authportal.dto.*;
import com.selva.authportal.exception.InvalidSubmissionException;
import com.selva.authportal.exception.ResourceNotFoundException;
import com.selva.authportal.exception.UserAlreadyExistsException;
import com.selva.authportal.model.*;
import com.selva.authportal.repository.SubmissionRepository;
import com.selva.authportal.repository.SubmissionSpecification;
import com.selva.authportal.repository.UserRepository;
import com.selva.authportal.repository.UserSpecification;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Service managing administrative operations (user provisioning, role management,
 * status toggling, submission oversight, and physical deletion).
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AdminService {

    private final UserRepository userRepository;
    private final SubmissionRepository submissionRepository;
    private final StorageService storageService;
    private final PasswordEncoder passwordEncoder;

    /**
     * Lists all system users matching optional search query and role filter.
     */
    @Transactional(readOnly = true)
    public List<UserResponse> getAllUsers(String query, Role role) {
        return userRepository.findAll(UserSpecification.withFilters(query, role)).stream()
                .sorted(Comparator.comparing(User::getCreatedAt).reversed())
                .map(UserResponse::fromEntity)
                .collect(Collectors.toList());
    }

    /**
     * Retrieves an individual user by ID.
     */
    @Transactional(readOnly = true)
    public UserResponse getUserById(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));
        return UserResponse.fromEntity(user);
    }

    /**
     * Creates a new user with an administrator-specified role.
     */
    @Transactional
    public UserResponse createUser(CreateUserRequest request) {
        String normalizedEmail = request.getEmail().trim().toLowerCase();

        if (userRepository.existsByEmail(normalizedEmail)) {
            throw new UserAlreadyExistsException("An account with email " + normalizedEmail + " already exists.");
        }

        String hashedPassword = passwordEncoder.encode(request.getPassword());

        User user = User.builder()
                .name(request.getName().trim())
                .email(normalizedEmail)
                .password(hashedPassword)
                .role(request.getRole())
                .authProvider(AuthProvider.LOCAL)
                .enabled(true)
                .build();

        User saved = userRepository.save(user);
        log.info("Admin created new user {} with role {}", normalizedEmail, request.getRole());
        return UserResponse.fromEntity(saved);
    }

    /**
     * Updates a user's system role.
     * Prevents demoting the final remaining administrator in the database.
     */
    @Transactional
    public UserResponse updateUserRole(Long userId, Role newRole, User currentAdmin) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));

        if (user.getRole() == Role.ADMIN && newRole != Role.ADMIN) {
            long adminCount = userRepository.countByRole(Role.ADMIN);
            if (adminCount <= 1) {
                throw new InvalidSubmissionException("Cannot change role: this is the final remaining administrator in the system.");
            }
        }

        user.setRole(newRole);
        User updated = userRepository.save(user);
        log.info("Admin {} updated role of user {} to {}", currentAdmin.getEmail(), user.getEmail(), newRole);
        return UserResponse.fromEntity(updated);
    }

    /**
     * Enables or disables a user account.
     * Prevents an administrator from disabling their own account or disabling the final administrator.
     */
    @Transactional
    public UserResponse toggleUserStatus(Long userId, boolean enabled, User currentAdmin) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));

        if (user.getId().equals(currentAdmin.getId()) && !enabled) {
            throw new InvalidSubmissionException("Administrators cannot disable their own active account.");
        }

        if (user.getRole() == Role.ADMIN && !enabled) {
            long adminCount = userRepository.countByRole(Role.ADMIN);
            if (adminCount <= 1) {
                throw new InvalidSubmissionException("Cannot disable the final remaining administrator in the system.");
            }
        }

        user.setEnabled(enabled);
        User updated = userRepository.save(user);
        log.info("Admin {} changed enabled status of user {} to {}", currentAdmin.getEmail(), user.getEmail(), enabled);
        return UserResponse.fromEntity(updated);
    }

    /**
     * Fetches all submissions across all students with optional filters for admin oversight.
     */
    @Transactional(readOnly = true)
    public List<SubmissionResponse> getAllSubmissions(Integer week, YearLevel year, Integer section, String studentId) {
        return submissionRepository.findAll(SubmissionSpecification.withFilters(week, year, section, studentId)).stream()
                .sorted(Comparator.comparing(Submission::getCreatedAt).reversed())
                .map(SubmissionResponse::fromEntity)
                .collect(Collectors.toList());
    }

    /**
     * Deletes a submission from the database and cleans up its physical directory on disk.
     */
    @Transactional
    public void deleteSubmission(Long submissionId) {
        Submission submission = submissionRepository.findById(submissionId)
                .orElseThrow(() -> new ResourceNotFoundException("Submission not found with id: " + submissionId));

        // Clean up files on disk
        try {
            Path submissionDir = storageService.resolveSubmissionDirectory(
                    submission.getWeek(),
                    submission.getSection(),
                    submission.getStudentId()
            );
            storageService.deleteDirectoryContents(submissionDir);
            Files.deleteIfExists(submissionDir);
        } catch (IOException e) {
            log.warn("Failed to clean up submission directory for id {}: {}", submissionId, e.getMessage());
        }

        submissionRepository.delete(submission);
        log.info("Deleted submission id {} for student {}", submissionId, submission.getStudentId());
    }

    /**
     * Aggregates key system metrics for the administrative dashboard.
     */
    @Transactional(readOnly = true)
    public AdminStatsResponse getSystemStats() {
        return AdminStatsResponse.builder()
                .totalUsers(userRepository.count())
                .studentCount(userRepository.countByRole(Role.STUDENT))
                .teacherCount(userRepository.countByRole(Role.TEACHER))
                .adminCount(userRepository.countByRole(Role.ADMIN))
                .totalSubmissions(submissionRepository.count())
                .build();
    }
}
